File: binarySearchTreeIterator/binaryTreeSearchIterator.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class BSTIterator{
public:
  BSTIterator(TreeNode *root) {
    storePath(root);
  }
  /** @return whether we have a next smallest number */
  bool hasNext(){
    return !s.empty();
  }
  /** @return the next smallest number */
  int next() {
    TreeNode *temp = s.top();
    s.pop();
    storePath(temp->right);
    return temp->val;
  }
  void storePath(TreeNode *t) {
    while (t) {
      s.push(t);
      t = t->left;
    }
  }
  stack<TreeNode *> s;
};

int main()
{
  int tree[7] = {5,3,6,2,4,9999,7};
  int size = sizeof(tree)/sizeof(tree[0]);
  vector<TreeNode *> vec;
  for (int i = 0; i < size; i++) {
    if (tree[i] != 9999) {
      vec.push_back(new TreeNode(tree[i]));
    } else {
      vec.push_back(NULL);
    }
  }
  for (int i = 0; i < size/2; i++) {
    if(!vec[i])
      continue;
    if (i*2 + 1 < size)
      vec[i]->left = vec[i*2 + 1];
    if (i*2 + 2 < size)
      vec[i]->right = vec[i*2 + 2];
  }
  BSTIterator i = BSTIterator(vec[0]);
  while (i.hasNext())
    cout << i.next() << " ";
  cout << endl;
  vec.clear();
  return 0;
}

File: 173.binarySearchTreeIterator/binaryTreeSearchIterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {

    public BSTIterator(TreeNode root) {
        s = new Stack<>();
        putLeftChildToStack(root);
    }
    
    public int next() {
        if (hasNext()) {
            TreeNode t = s.pop();
            if (t.right != null) {
                putLeftChildToStack(t.right);
            }
            return t.val;
        }
        return -1;
    }
    
    public boolean hasNext() {
        return !s.isEmpty();
    }
    
    private void putLeftChildToStack(TreeNode t) {
        while (t != null) {
            s.push(t);
            t = t.left;
        }
    }
    
    Stack<TreeNode> s;
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */
